### PR TITLE
Update to new types from rust-secp-zkp

### PR DESCRIPTION
**Pull request by sanket1729**
First part of #73 . Implementation of pset

### CODE REVIEW

File: src/pset/map/global.rs [hunk 1]
```diff
@@ -236,13 +244,23 @@ impl Map for Global {
 
         // Serializing version only for non-default value; otherwise test vectors fail
         if self.version > 0 {
+            let ver = self.version; //hack to use macro
+            impl_pset_get_pair!(
+                rv.push_mandatory(ver as <PSET_GLOBAL_VERSION, _>)
+            )
+        }
```
Comment (apoelstra) [inline src/pset/map/global.rs:251]: Now that this is a mandatory field, we should serialize it unconditionally I think.

File: src/pset/map/global.rs [hunk 2]
```diff
@@ -426,14 +473,21 @@ impl Decodable for Global {
...
-            let mut rv: Global = Global::from_unsigned_tx(tx)?;
-            rv.version = version.unwrap_or(0);
-            rv.xpub = xpub_map;
-            rv.unknown = unknowns;
-            Ok(rv)
-        } else {
-            Err(Error::MustHaveUnsignedTx.into())
-        }
+        // Mandatory fields
+        let version = version.ok_or(Error::MissingPsetVersion)?;
```
Comment (apoelstra) [inline src/pset/map/global.rs:477]: We should require the version to be 2, not just that it be present.

File: src/pset/mod.rs [hunk 3]
```diff
@@ -180,19 +178,19 @@ impl PartiallySignedTransaction {
...
         let mut outputs = vec![];
 
-        for psetin in self.inputs.into_iter() {
+        for psetin in self.inputs.iter() {
             let txin = TxIn {
                 previous_output: OutPoint::new(psetin.previous_txid, psetin.previous_output_index),
                 is_pegin: psetin.previous_output_index & (1 << 30) != 0,
                 has_issuance: psetin.previous_output_index & (1 << 31) != 0,
-                script_sig: psetin.final_script_sig.unwrap_or(Script::new()),
+                script_sig: psetin.final_script_sig.clone().unwrap_or(Script::new()),
```
Comment (apoelstra) [inline src/pset/mod.rs:193]: I know this is the same as the old code, but FYI we could use `unwrap_or_default` here (and a few other places)

**Comment by apoelstra** [comment]
This looks great!!

Can you add a new `deserialize_pset` fuzztest target with

```
fn do_test(data: &[u8]) {
    let psbt: Result<elements::pset::PartiallySignedTransaction, _> = elements::encode::deserialize(data);
    match psbt {
        Err(_) => {}, 
        Ok(psbt) => {
            let ser = elements::encode::serialize(&psbt);
            let deser: elements::pset::PartiallySignedTransaction = elements::encode::deserialize(&ser).unwrap();
            // Since the fuzz data could order psbt fields differently, we compare to our deser/ser instead of data
            assert_eq!(ser, elements::encode::serialize(&deser));
        }   
    }   
}
```

? This triggers right now if you try to parse a PSET with version 0 but it should be fixed by requiring the version to be 2.

File: src/pset/map/global.rs (hunk 1 above)
Comment (sanket1729) [reply src/pset/map/global.rs:251]: It's only a mandatory field in v2

File: src/pset/map/global.rs (hunk 2 above)
Comment (sanket1729) [reply src/pset/map/global.rs:477]: Sure, will fix this. I was thinking about future versions.

**Comment by sanket1729** [comment]
I think this is expected behavior based on spec for v0. But as you say fixing this field to 2 solves all the issues.

**Comment by sanket1729** [comment]
@apoelstra made the requested changes and added the fuzz tests.

File: src/pset/map/global.rs (hunk 1 above)
Comment (apoelstra) [reply src/pset/map/global.rs]: But there are no other versions of PSET.

**Review by apoelstra — APPROVED** [review]
ack a8397ef481aa0a348cffaf3d19b918f6263f93f3

**Pull request merged**